llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -177,6 +182,81 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineSetExtract(
   return true;
 }
 
+// Returns a map with InsertIndices and registers holding the insert values.
+std::optional<std::map<unsigned, Register>>
+AIE2PreLegalizerCombinerImpl::getVectorInsertIndices(
+    MachineInstr *CurMI, unsigned SclSrcBits, MachineRegisterInfo &MRI) const {
+  std::map<unsigned, Register> RegMap;
+  auto Is8BitVInsert = [](const MachineInstr *MI) {
+    return isa<GIntrinsic>(MI) && cast<GIntrinsic>(*MI).getIntrinsicID() ==
+                                      Intrinsic::aie2_vinsert8_I512;
+  };
+  auto IsSet = [](const MachineInstr *MI) {
+    return isa<GIntrinsic>(MI) && (cast<GIntrinsic>(*MI).getIntrinsicID() ==
+                                       Intrinsic::aie2_set_I512_I128 ||
+                                   cast<GIntrinsic>(*MI).getIntrinsicID() ==
+                                       Intrinsic::aie2_set_I512_I256);
+  };
+
+  while (Is8BitVInsert(CurMI)) {
+    // In this case of G_INTRINSIC operand 1 is target intrinsic
+    const Register SrcReg = CurMI->getOperand(2).getReg();
+    const Register IdxReg = CurMI->getOperand(3).getReg();
+    const Register SclSrcReg = CurMI->getOperand(4).getReg();
+
+    // Collecting registers and their indices
+    auto Cst = getIConstantVRegValWithLookThrough(IdxReg, MRI);
+    if (!Cst ||
+        !RegMap.try_emplace(Cst->Value.getZExtValue(), SclSrcReg).second)
+      break;
+    CurMI = getDefIgnoringCopies(SrcReg, MRI);
+
+    // Combining Set and Extract to fetch next VInsert
+    if (IsSet(CurMI) && tryToCombineSetExtract(*CurMI))
+      CurMI = getDefIgnoringCopies(SrcReg, MRI);
```

Comment:
Isn't `tryToCombineSetExtract` potentially deleting `CurMI`? Is it always safe to call `CurMI = getDefIgnoringCopies(SrcReg, MRI);`?